### PR TITLE
Start migrating CircleCI workflow to GitHub actions

### PR DESCRIPTION
**Pull request by owais**
Introduce GitHub Actions build/test/publish workflow. This ports most of the build-publish workflow from CircleCI to Github Actions. Most of the pipeline and the scripts have been taken from the core repo and adapted where necessary to work with contrib.

### Pending work items (will be completed in separate PRs):

- Enable publish-dev and publish-stable jobs (disbaled for now; need to setup secrets with proper protection).
- Integrate github issuegenerator to report some failures such as loadtest as github issues.
- Disable CircleCI.

#### Duplication in Github YAML (separate work item):
Github does not support re-usable steps, running actions from within other actions or YAML anchors like CircleCI. As a result there is a lot of duplication in the file. I'll follow up this with a solution to elimination the duplication. I'm thinking to have a `build-and-test.in.yaml` file which will use anchors to re-use blocks (like go and tooling cache, publish stable+dev, etc) and then have a tool that generates another `build-and-test.yaml` from that file with expanded anchors. This is a separate work item though so not included in this PR.

#### Can't re-run only failed jobs
This can be extremely annoying. If a single job fails in a workflow, there is no way to retry just the failed job. Only way to retry is to retry the entire workflow. This has been very annoying for personally with other Otel and Splunk projects. GitHub has acknowledged that the feature is on their roadmap but there is no ETA on this. We can hack together a solution that stores the result (pass/fail) of every job in workflow somewhere and on re-runs skips jobs if:

- they passed in the previous run of the same workflow
- no files affecting the job had any changes

This is very much possible to implement but not sure if it is worth it. Either way we can invest time into it IF it turns out to be a big issue.

#### CircleCI
image

#### Github Actions
image

GitHub workflow lists all jobs even if they're not supposed to run in that workflow such as publish-dev/publish-stable in ^ screenhot.

### CODE REVIEW

**Comment by mx-psi** [comment]
One item that IMO we should add to the pending work items list is to enable Dependabot for Github Actions. Maybe we should have a separate issue to track this?

**Comment by owais** [comment]
Yes, that doesn't seem related to CI jobs. Separate issue would be better IMO.

**Comment by tigrannajaryan** [comment]
> I'm thinking to have a `build-and-test.in.yaml` file which will use anchors to re-use blocks (like go and tooling cache, publish stable+dev, etc) and then have a tool that generates another `build-and-test.yaml` from that file with expanded anchors. This is a separate work item though so not included in this PR.

SGTM.

> #### Can't re-run only failed jobs

A good reason for us to focus on making our tests rock stable (the tests are the offenders which require rerunning). Let's spend the time we would spend on finding a solution to rerun single jobs on making our tests stable. It is a better investment IMO.

**Review by tigrannajaryan — COMMENTED** [review]
The approach LGTM, thanks for working on this.

**Comment by owais** [comment]
> A good reason for us to focus on making our tests rock stable (the tests are the offenders which require rerunning). Let's spend the time we would spend on finding a solution to rerun single jobs on making our tests stable. It is a better investment IMO.

Definitely but despite that this can be an annoyance for people working on new features and using CI for tests. Luckily I think collector's local workflow is simple enough to run and verify everything locally so CI shouldn't be required to be part of the local iteration process.

**Comment by tigrannajaryan** [comment]
@owais can you please rebase, the failed loadtest should pass now.

File: .github/workflows/build-and-test.yml
Comment (tigrannajaryan) [inline .github/workflows/build-and-test.yml:532]: Can you add a comment to explain why this is excluded?

File: .github/workflows/build-and-test.yml
Comment (tigrannajaryan) [inline .github/workflows/build-and-test.yml:533]: Do we need this? Can we delete commented code or turn it into a TODO comment?

File: .github/workflows/build-and-test.yml
Comment (tigrannajaryan) [inline .github/workflows/build-and-test.yml:479]: Same here: Can we delete commented code or turn it into a TODO comment?

**Review by tigrannajaryan — COMMENTED** [review]
LGTM, just one comment left, will merge after that is addressed.

File: .github/workflows/build-and-test.yml
Comment (owais) [reply .github/workflows/build-and-test.yml:480]: Done

**Comment by owais** [comment]
Looks like loadtests are unreliable on GH actions compared to Circle. Nothing blocks this PR but we might want to keep loadtests on Circle for a while. In the long run, we probably need funding to be able to run loadtests in the cloud where we can predict performance a bit more reliably.

**Review by gramidt — CHANGES_REQUESTED** [review]
There appears to be a new empty file 'otelcontribcol-all-sys' as part of this PR. Was this intentional?

**Comment by owais** [comment]
@gramidt No, removed. Thanks

**Review by tigrannajaryan — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@bogdandrutu this looks good to me. We can merge if you don't have any additional thoughts.

**Review by gramidt — APPROVED** [review]

File: internal/buildscripts/packaging/msi/make.ps1
Comment (bogdandrutu) [inline internal/buildscripts/packaging/msi/make.ps1:80]: New line.

**Review by bogdandrutu — APPROVED** [review]
This looks good to me as a high-level direction, did not look at the code but trust @tigrannajaryan

**Pull request merged**
